src/dist/failure_detector/failure_detector.cpp

```diff
@@ -228,13 +225,17 @@ void failure_detector::process_all_records()
                 report(record.node, true, false);
             }
         }
-    }
 
-    if ( expire.size() > 0 )
-    {
-        on_master_disconnected(expire);
+        /*
+         * The disconnected event MUST be under the protection of the _lock
+         */
+        if ( expire.size() > 0 )
+        {
+            on_master_disconnected(expire);
+        }
```

Comment:
on_master_disconnected() is protected by lock, but why the following on_worker_disconnected() is not protected by lock?